### DISCUSSION TITLE
git references in Dockerfile ADD instruction

### DISCUSSION THREAD

**Discussion by morremeyer** [renovatebot/renovate#29573] [Suggest an Idea]
## Description

The Dockerfile `ADD` instruction supports git references, for example

```
ADD https://github.com/moby/buildkit.git#v0.10.1 /buildkit
```

This is documented in https://docs.docker.com/reference/dockerfile/#add---keep-git-dir.

## Proposal

Add support in renovate to bump the references after the `#`. 

**Notes**:

* Needs to ignore all options to the `ADD` command

### Datasource detection

How can the best datasource be detected?

In the case of a tag or release, it could be the respective `*-tags` or `*-releases` datasource, in the case of a commit ID, there is no sensible default since the branch it refers to is not clear.

_Idea_: Detect if it is a commit ID, if yes, ignore, if no, default to `*-tags` or `*-releases` datasource?

**Comment by rarkins** [comment]
Did you mean to remove out `--keep-git-dir` from your example, or was it a typo?

**Reply by morremeyer** [reply]
I meant to remove it to simplify the example. The actual implementation will have to ignore all options to the `ADD` command, which is a good point. I added that to the initial comment.

**Reply by rarkins** [reply]
But I think that the flags are essential for identifying that it's a git reference

**Reply by morremeyer** [reply]
git references do not neccessarily require a flag, e.g.

```Dockerfile
ADD https://github.com/conda/conda-index.git#0.5.0 /app/conda-index
```

is also valid.

I would detect git references that with the `.git` suffix.

**Reply by rarkins** [reply]
I've done a quick search plus scan of that docs page and I haven't found an authoritative reference to the formats of git repo which Docker supports for `ADD`.

We could take the easy way out and just support https://github.com initially, but ideally we'd have full `git-refs` support.

I wouldn't object to a PR adding the example syntax you gave though (i.e. simple github.com repo with a semver ref)

**Reply by morremeyer** [reply]
I have had the same problem, I would've expected a link to a format documentation somewhere. I'll take a deeper look if I can find that if I get to implementing it.

I like the idea of starting simple, just supporting GitHub + semver refs.

I'm putting this on my to do list, we'll see when I get to it.
